### ISSUE TITLE
Regression with inference of type args through decorator on class constructor

### ISSUE THREAD

**Issue by adampauls** [microsoft/pyright#5396]
**Describe the bug**
Minimization (sort of) of a bug first mentioned in https://github.com/microsoft/pyright/issues/5369. 

**To Reproduce**
See code below. If a type argument is explicitly specified on the commented line, no error is issued. The issue appears as of command line pyright 1.1.315 and persists on 1.1.316 works fine on 1.1.314 or below. 

**Expected behavior**
The code should typecheck whether or not the type argument is specified. 

**Code or Screenshots**
```python
Slf = TypeVar("Slf")
Params = ParamSpec("Params")
Return = TypeVar("Return")

class MethodModifier(Protocol[Slf, Params, Return]):
    def __call__(self, func: Callable[Concatenate[Slf, Params], Return]) -> Callable[Concatenate[Slf, Params], Return]:
        ...

T1 = TypeVar("T1")
T2 = TypeVar("T2")

Params2 = ParamSpec("Params2")

def my_decorator(class_factory: Callable[Params2, Any]) -> MethodModifier[Any, Params2, Any]:
    ...


def test_inference() -> None:
    @dataclass(frozen=True)
    class TestClass(Generic[T1]):
        generic: T1

    class TestDecoratedClass(Generic[T2]):
        @my_decorator(TestClass) # works fine with TestClass[T2]
        def takes_type_var(self, generic: T2) -> T2:
            return generic
```

Error is:

```
error: Argument of type "(self: Self@TestDecoratedClass[T2@TestDecoratedClass], generic: T2@TestDecoratedClass) -> T2@TestDecoratedClass" cannot be assigned to parameter "func" of type "(Any, generic: T1@TestClass) -> Any" in function "__call__"
    Type "(self: Self@TestDecoratedClass[T2@TestDecoratedClass], generic: T2@TestDecoratedClass) -> T2@TestDecoratedClass" cannot be assigned to type "(Any, generic: T1@TestClass) -> Any"
      Parameter 2: type "T1@TestClass" cannot be assigned to type "T2@TestDecoratedClass"
        Type "T1@TestClass" cannot be assigned to type "T2@TestDecoratedClass" (reportGeneralTypeIssues)
```

**VS Code extension or command-line**
Appears as of command line pyright 1.1.315, works fine on 1.1.314 or below.

**Comment by erictraut** [comment]
This change was made to address bug #5328. Reverting the change will make this issue go away, but it will break other legitimate cases. I think the current behavior is arguably more correct than the previous behavior. It's also consistent with mypy, which emits the same error for your code sample. For that reason, I don't consider this a regression or a bug. It's the new intended behavior.

To work around this, you should provide a type argument for `TestClass` when passing it as an argument to the `my_decorator` function. I'm guessing that the correct expression is `TestClass[T2]`, but perhaps `TestClass[Any]` is your intent.

**Comment by adampauls** [comment]
Thanks for the quick response as always. I can't say I follow. I agree that the fix you made should stay, but it still seems like a new bug was introduced. The error message alone indicate a bug to me -- why can T1 not be assigned to T2? They are both free type variables?

It seems that before your fix, type vars were specialized to soon, but now they are being specialized too late? Perhaps it is more work than it is worth to fix, but the error message makes it look to me like the type checker has figured out that T1 must be T2 so maybe there's a one-liner somewhere. 

In any case, I wouldn't mind adding the explicit type args if it weren't for the runtime bug mentioned in the other issue, but that's not pyright's fault!
